File: leetcode/253.py
"""
---
title: Meeting Rooms II
difficulty: medium
level: 2
links: 
- https://leetcode.com/problems/meeting-rooms-ii
---
Given an array of meeting time intervals consisting of start and end times
[[s1,e1],[s2,e2],...] (si < ei), find the minimum number of conference rooms
required.

Example 1:

Input: [[0, 30],[5, 10],[15, 20]]
Output: 2

Example 2:

Input: [[7,10],[2,4]]
Output: 1
"""
from typing import List
import heapq


def minMeetingRooms(intervals: List[List[int]]) -> int:
    # minimally we need to sort by start time, because we can't process this
    # in just any order. these are chronological events, and we need to create
    # a timeline to allow us to build up the # of required rooms

    # an end time signifies a room freeing up
    # a start time signifies that we require a room, but not necessarily a new room

    # as we process each individual start time, we need to consider the next
    # soonest end time for a meeting. if the next soonest end time for a meeting
    # is after the start time we're currently looking at, we need to increment
    # the number of rooms required, if not, the number of rooms required stays the
    # same

    # i will intially sort on start time, then create a priority queue
    # based on end time, so that the next soonest ending meeting will
    # be on top. you could also seperate start and end time into two arrays
    # and sort them seperately with the same effect (both O(n log n))

    rooms: int = 0

    # reverse sort on start time in place, so we can cheaply pop
    intervals.sort(key=lambda time: -time[0])

    # make a copy here, swapping end and start so we can use a py min heap
    ends: List[List[int]] = [[time[1], time[0]] for time in intervals]

    # min heap the end times
    heapq.heapify(ends)

    while intervals:
        start, _ = intervals.pop()

        # peek the next end time, if it hasn't ended yet, start up a room
        if start < ends[0][0]:
            rooms += 1
        else:
            heapq.heappop(ends)

    return rooms


result = minMeetingRooms([[15, 20], [0, 30], [5, 10]])
print(result)
assert result == 2

result = minMeetingRooms([[7, 10], [2, 4]])
print(result)
assert result == 1

File: season2/253.js
/**
 * @param {number[][]} intervals
 * @return {number}
 */
const minMeetingRooms = (intervals) => {
  // sort the array
  // turn each interval into two events, one is start, one is end
  // iterate over the new events array
  // every time we encounter a start -- increment the value of rooms required
  // every time we encounter any end -- decrement the rooms required
  // O(n * log(n))
  
  const events = [];
  
  for (const i of intervals) {
    events.push({type:'start', time: i[0]});
    events.push({type:'end', time: i[1]});
  }
  
  events.sort((a, b) => {
    const n = a.time - b.time;
    
    if (n === 0) {
      if (a.type === "end") {
        return -1;
      } else {
        return 1;
      }
    }
    
    return n;
  });
  
  let max = 0, rooms = 0;
  
  for (const e of events) {
    if (e.type === "start") {
      rooms += 1;
      max = Math.max(rooms, max);
    } else {
      rooms -= 1;
    }
  }
  
  return max;
};
